pkg/operator/etcdcertsigner/etcdcertsignercontroller.go

```diff
@@ -331,6 +343,49 @@ func (c *EtcdCertSignerController) createSecretForNode(node *corev1.Node) error
 	return nil
 }
 
+func createPeerDNSSubjectAlternativeNamesFromMemberList(etcdDiscoveryDomain string, etcdMembers []*etcdserverpb.Member, nodeInternalIPs []string) []string {
+	var found bool
+	var errs []string
+	hosts := []string{"localhost"}
+	for _, m := range etcdMembers {
+		// 4.3 naming
+		if strings.HasPrefix(m.Name, "etcd-member") {
+			found = true
+		}
+		if len(m.PeerURLs) > 0 && len(m.ClientURLs) > 0 {
+			peer, err := url.Parse(m.PeerURLs[0])
+			if err != nil {
+				errs = append(errs, err.Error())
+				continue
+			}
+			if addr := net.ParseIP(peer.Hostname()); addr != nil {
+				continue
+			}
+			client, err := url.Parse(m.ClientURLs[0])
+			if err != nil {
+				errs = append(errs, err.Error())
+				continue
+			}
+			for _, nodeIP := range nodeInternalIPs {
+				if nodeIP == client.Hostname() {
+					hosts = append(hosts, peer.Hostname())
+				}
+			}
+		}
+	}
+
+	switch {
+	case len(errs) > 0:
+		break // fallback to wildcard
+	case len(hosts) == 1 && found:
+		break // fallback to wildcard
+	case len(hosts) > 1 || !found:
+		return append(hosts, nodeInternalIPs...)
```

Comment:
just add all the things all the time, so you don't have branches. Then in 4.5, just stop adding the 4.3 compatibility ones.